columnphysics/icepack_wavefracspec.F90

```diff
@@ -525,38 +531,66 @@ spec_elambda,             & ! spectrum as a function of wavelength (m^-1 s^-1)
             summand = spec_coeff*COS(2*pi*X(j)/lambda+phi)
             eta(j)  = SUM(summand)
          end do
- 
+
+         fraclengths(:) = c0 
          if ((SUM(ABS(eta)) > puny).and.(hbar > puny)) then 
-            call get_fraclengths(X, eta, fraclengths, hbar, e_stop)
+            call get_fraclengths(X, eta, fraclengths, hbar)
          end if
-      end do
- 
-      frachistogram(:) = c0
-
-      if (.not. e_stop) then
 
          ! convert from diameter to radii
          fraclengths(:) = fraclengths(:)/c2
 
-         ! bin into FS cats
-         ! highest cat cannot be fractured into
-         do j = 1, size(fraclengths)
+ 
+         if (ALL(fraclengths.lt.floe_rad_l(1))) then
+             frac_local(:) = c0
+         else
+            frachistogram(:) = c0
+            allfraclengths((iter-1)*nx+1:(iter)*nx) = fraclengths(1:nx)
+
+            ! bin into FS cats
+            do j = 1, size(fraclengths)
+            if (allfraclengths(j).gt.floe_rad_l(1)) then
             do k = 1, nfsd-1
-               if ((fraclengths(j) >= floe_rad_l(k)) .and. &
-                   (fraclengths(j) < floe_rad_l(k+1))) then
+               if ((allfraclengths(j) >= floe_rad_l(k)) .and. &
+                   (allfraclengths(j) < floe_rad_l(k+1))) then
                   frachistogram(k) = frachistogram(k) + 1
                end if
             end do
+            if (allfraclengths(j)>floe_rad_l(nfsd)) frachistogram(nfsd) = frachistogram(nfsd) + 1
+            end if
+            end do
 
-         end do
-      end if
+            do k = 1, nfsd
+             frac_local(k) = floe_rad_c(k)*frachistogram(k)
+            end do
 
-      do k = 1, nfsd
-         frac_local(k) = floe_rad_c(k)*frachistogram(k)
-      end do
+            ! normalize
+            if (SUM(frac_local) /= c0) frac_local(:) = frac_local(:) / SUM(frac_local(:))
+
+         end if 
+
+         ! wave fracture run to convergence
+         if (trim(wave_spec_type).eq.'random') then
+
+             ! check avg frac local against previous iteration
+             fracerror = SUM(ABS(frac_local - prev_frac_local))/nfsd
+
+             if (fracerror.lt.errortol) EXIT
 
-      ! normalize
-      if (SUM(frac_local) /= c0) frac_local(:) = frac_local(:) / SUM(frac_local(:))
+             if (iter.gt.100) then
+               print *, 'fracerror ',fracerror
+               print *, 'before ',prev_frac_local
+               print *, 'after ',frac_local
+               stop 'wave_frac did not converge'
```

Comment:
We need to use the warning software to do this abort (and check for others in the FSD code, like this).  @apcraig could you help @lettie-roach with this?